src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,174 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+    // (grace time to allow next search to start)
     // Return Value:
     // - <none>
-    void TermControl::_Search(const winrt::hstring& text, const bool goForward, const bool caseSensitive)
+    fire_and_forget TermControl::_SearchAsync(std::optional<bool> goForward, Windows::Foundation::TimeSpan const& delay)
     {
-        if (text.size() == 0 || _closing)
+        // Run only if the live search state was initialized
+        if (_closing || !_searchState.has_value())
         {
             return;
         }
 
-        const Search::Direction direction = goForward ?
-                                                Search::Direction::Forward :
-                                                Search::Direction::Backward;
+        const auto originalSearchId = _searchState.value().SearchId;
+        auto weakThis{ this->get_weak() };
 
-        const Search::Sensitivity sensitivity = caseSensitive ?
-                                                    Search::Sensitivity::CaseSensitive :
-                                                    Search::Sensitivity::CaseInsensitive;
+        // If no matches were computed it means we need to perform the search
+        if (!_searchState.value().Matches.has_value())
+        {
+            // Before we search, let's wait a bit:
+            // probably the search criteria or the data are still modified.
+            co_await winrt::resume_after(delay);
 
-        Search search(*GetUiaData(), text.c_str(), direction, sensitivity);
-        auto lock = _terminal->LockForWriting();
-        if (search.FindNext())
+            // Switch back to Dispatcher so we can set the Searching status
+            co_await winrt::resume_foreground(Dispatcher());
+            if (auto control{ weakThis.get() })
+            {
+                // If search box was collapsed or the new one search was triggered - let's cancel this one
+                if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                {
+                    co_return;
+                }
+
+                // Let's mark the start of searching
+                if (_searchBox)
+                {
+                    _searchBox->SetStatus(_searchState.value().Status());
+                    _searchBox->SetNavigationEnabled(false);
+                }
+
+                std::vector<std::pair<COORD, COORD>> matches;
+                if (!_searchState.value().Text.empty())
+                {
+                    const Search::Sensitivity sensitivity = _searchState.value().CaseSensitive ?
+                                                                Search::Sensitivity::CaseSensitive :
+                                                                Search::Sensitivity::CaseInsensitive;
+
+                    Search search(*GetUiaData(), _searchState.value().Text.c_str(), Search::Direction::Forward, sensitivity);
+                    while (co_await _SearchOne(search))

```

Comment:
/note this while loop collects up all the search matches, on a BG thread. If a new search is started, then it stops collecting matches

Reply:
oh but _SearchOne returns to the FG thread. Hmm. Does it need to do that? Could we do the whole search on one pass on the BG thread, then hop back to finish the UI bits?

Reply:
@zadjii-msft - yeah.. I switched to the foreground thread to avoid synchronization. An alternative is to do everything in BG and synchronize the state